src/http-api/resources/files.js

```diff
@@ -33,18 +40,152 @@ exports.cat = {
   // main route handler which is called after the above `parseArgs`, but only if the args were valid
   handler: (request, reply) => {
     const key = request.pre.args.key
+    const ipfs = request.server.app.ipfs
 
-    request.server.app.ipfs.files.cat(key, (err, stream) => {
+    ipfs.files.cat(key, (err, stream) => {
       if (err) {
         log.error(err)
         return reply({
           Message: 'Failed to cat file: ' + err,
           Code: 0
         }).code(500)
       }
-      stream.on('data', (data) => {
-        return reply(data)
+
+      // hapi is not very clever and throws if no
+      // - _read method
+      // - _readableState object
+      // are there :(
+      stream._read = () => {}
+      stream._readableState = {}
+      return reply(stream).header('X-Stream-Output', '1')
+    })
+  }
+}
+
+exports.get = {
+  // uses common parseKey method that returns a `key`
+  parseArgs: exports.parseKey,
+
+  // main route handler which is called after the above `parseArgs`, but only if the args were valid
+  handler: (request, reply) => {
+    const key = request.pre.args.key
+    const ipfs = request.server.app.ipfs
+    const pack = tar.pack()
+
+    ipfs.files.getPull(key, (err, stream) => {
+      if (err) {
+        log.error(err)
+
+        reply({
+          Message: 'Failed to get file: ' + err,
+          Code: 0
+        }).code(500)
+        return
+      }
+
+      pull(
+        stream,
+        pull.asyncMap((file, cb) => {
+          const header = {name: file.path}
+
+          if (!file.content) {
+            header.type = 'directory'
+            pack.entry(header)
+            cb()
+          } else {
+            header.size = file.size
+            toStream.source(file.content)
+              .pipe(pack.entry(header, cb))
+          }
+        }),
+        pull.onEnd((err) => {
+          if (err) {
+            log.error(err)
+
+            reply({
+              Message: 'Failed to get file: ' + err,
+              Code: 0
+            }).code(500)
+            return
+          }
+
+          pack.finalize()
+        })
+      )
+
+      // the reply must read the tar stream,
+      // to pull values through
+      reply(pack).header('X-Stream-Output', '1')
+    })
+  }
+}
+
+exports.add = {
+  handler: (request, reply) => {
+    if (!request.payload) {
+      return reply('Array, Buffer, or String is required.').code(400).takeover()
+    }
+
+    const ipfs = request.server.app.ipfs
+    // TODO: make pull-multipart
```

Comment:
Track this in a issue. @xicombd you probably would like to get that working, since you did the first multipart parser :)